### PR TITLE
Added new versions of StringManipulation .java files

### PR DESCRIPTION
**Pull request by AaronBird759**

### CODE REVIEW

File: src/test/java/StringManipulationTest.java [hunk 1]
```diff
@@ -26,19 +25,29 @@ public void testCount1() {
...
+        int length = manipulatedstring.count();
+        assertEquals(11, length);
     }
 
+    // another test, but this time we add quotation marks to see if it messes with
+    // word count
     @Test
     public void testCount4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("This a string that can't fail just because \"can't\" is in the string");
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:48]: This test seems redundant with Test case #1. Can it be replaced with testing extra whitespace (EX: " abc") and the one where string has more than 1 whitespace between words (EX: "abc  def")?

File: src/test/java/StringManipulationTest.java [hunk 2]
```diff
@@ -50,100 +59,152 @@ public void testRemoveNthCharacter1() {
...
     public void testGeSubStrings1() {
         manipulatedstring.setString("This is my string");
-        String [] sStings = manipulatedstring.getSubStrings(3, 4);
+        String[] sStings = manipulatedstring.getSubStrings(3, 4);
 
         assertEquals(sStings[0], "my");
         assertEquals(sStings[1], "string");
     }
 
+    // test to make sure that it can scale to more substrings
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:114]: Recommend modifying this test case so that it calls getSubStrings(1,15) (i.e. get all the words in a string).

File: src/test/java/StringManipulationTest.java [hunk 3]
```diff
@@ -50,100 +59,152 @@ public void testRemoveNthCharacter1() {
...
+    // IllegalArgumentException
     @Test
     public void testGeSubStrings3() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("This Test should throw IllegalArgumentException for startWord <= 0");
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.getSubStrings(0, 5));
     }
+
+    // tests thats endWord cannot be <= 0 and should throw a
+    // IllegalArgumentException
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:137]: Modify testGeSubStrings4 so that startWord is an integer greater than 0. This ensures that the exception is caused by the endWord <= 0, not because startWord is <= 0.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,94 @@
 public class StringManipulation implements StringManipulationInterface {
```
Comment (Emc2Ikeda) [inline src/main/java/StringManipulation.java:2]: Modify or add additional test cases in StringManipulationTest.java. If new test cases failed, modify this file so that it passes the said test cases.

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -2,7 +2,6 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:4]: Recommend adding some test cases for setString() and getString() for sake of sanity. Be sure to add cases when this.string is null and when you try to set this.string to null as well as when string is empty.

File: src/test/java/StringManipulationTest.java [hunk 6]
```diff
@@ -50,100 +59,152 @@ public void testRemoveNthCharacter1() {
...
-    public void testRestoreString5()
-    {
-        fail("Not yet implemented");
-
+    public void testRestoreString5() {
+        manipulatedstring.setString("left");
+        int[] array;
+        array = new int[] { 0, 1, 2, 3, 4, 5 };
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.restoreString(array));
     }
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:209]: Add another test case for restoreString() for case when string contains more than 1 word, such as "Hello world".

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -2,7 +2,6 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:5]: Also add cases in each method when string is null or empty.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -50,100 +59,152 @@ public void testRemoveNthCharacter1() {
...
     }
+
+    // tests the case where string length is shorter than endWord and throws a
+    // IndexOutOfBoundsException if it is
     @Test
     public void testGeSubStrings6() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("This Test should throw IndexOutOfBoundsException for endWord > string length");
+        assertThrows(IndexOutOfBoundsException.class, () -> manipulatedstring.getSubStrings(1, 100));
     }
```
Comment (Emc2Ikeda) [inline src/test/java/StringManipulationTest.java:159]: Add test case for getSubStrings() when startWord = endWord.

**Review by Emc2Ikeda — CHANGES_REQUESTED** [review]
Watch out for corner cases not described in the interface. For example, how should methods act when this.string is null or empty? I noted the ones I found in inline comments for StringManipulationTest.java. Then, re-test the StringManipulation class. If the new test cases fail, modify StringManipulation so that it passes all the new tests.
